File: src/xyang/validator/if_feature_eval.py
"""
Evaluate YANG 1.1 ``if-feature`` boolean expressions (RFC 7950 §7.20.2, §14).

Multiple ``if-feature`` substatements on one node are combined with logical AND.
"""

from __future__ import annotations

from typing import AbstractSet, Dict, List, Mapping, Optional, Set

from ..module import YangModule


class IfFeatureEvalError(Exception):
    """Malformed ``if-feature`` expression."""


def reachable_modules(root: YangModule) -> List[YangModule]:
    """``root`` and all modules reachable via ``import`` (transitive)."""
    out: List[YangModule] = []
    seen: Set[int] = set()

    def walk(m: YangModule) -> None:
        if id(m) in seen:
            return
        seen.add(id(m))
        out.append(m)
        for im in m.import_prefixes.values():
            walk(im)

    walk(root)
    return out


def build_enabled_features_map(
    root: YangModule,
    override: Optional[Mapping[str, AbstractSet[str]]],
) -> Dict[str, AbstractSet[str]]:
    """
    For each reachable module, the set of feature names that are *enabled*.

    * ``override is None``: every *declared* feature is enabled (backward compatible).
    * ``override[m]`` present: only those feature names are enabled for module ``m``.
    * Module not listed in ``override``: all declared features enabled.
    """
    modules = reachable_modules(root)
    result: Dict[str, AbstractSet[str]] = {}
    for m in modules:
        if override is None or m.name not in override:
            result[m.name] = frozenset(m.features)
        else:
            result[m.name] = frozenset(override[m.name])
    return _prune_per_feature_if_features(modules, result)


def feature_is_supported(
    ctx_module: YangModule,
    enabled_by_module: Mapping[str, AbstractSet[str]],
    ref: str,
) -> bool:
    """
    Whether feature reference *ref* (``name`` or ``prefix:name``) is supported
    in the context of *ctx_module* (the module where the ``if-feature`` appears).
    """
    if ":" in ref:
        pref, _, fname = ref.partition(":")
        mod = ctx_module.resolve_prefixed_module(pref)
        if mod is None:
            return False
    else:
        mod = ctx_module
        fname = ref
    if fname not in mod.features:
        return False
    enabled = enabled_by_module.get(mod.name)
    if enabled is None:
        return False
    return fname in enabled


def _tokenize(expr: str) -> List[str]:
    tokens: List[str] = []
    i = 0
    n = len(expr)
    while i < n:
        c = expr[i]
        if c.isspace():
            i += 1
            continue
        if c in "()":
            tokens.append(c)
            i += 1
            continue
        j = i
        while j < n and not expr[j].isspace() and expr[j] not in "()":
            j += 1
        tokens.append(expr[i:j])
        i = j
    return tokens


class _IfFeatureParser:
    __slots__ = ("_ctx", "_enabled", "_i", "_toks")

    def __init__(
        self,
        tokens: List[str],
        ctx_module: YangModule,
        enabled_by_module: Mapping[str, AbstractSet[str]],
    ) -> None:
        self._toks = tokens
        self._i = 0
        self._ctx = ctx_module
        self._enabled = enabled_by_module

    def _peek(self) -> Optional[str]:
        return self._toks[self._i] if self._i < len(self._toks) else None

    def _eat(self, expected: Optional[str] = None) -> str:
        t = self._peek()
        if t is None:
            raise IfFeatureEvalError("unexpected end of expression")
        if expected is not None and t != expected:
            raise IfFeatureEvalError(f"expected {expected!r}, got {t!r}")
        self._i += 1
        return t

    def parse_expr(self) -> bool:
        left = self.parse_term()
        while self._peek() == "or":
            self._eat("or")
            # Parse RHS unconditionally — Python ``or`` short-circuits and would skip
            # consuming tokens, breaking ``at_end()`` and wrong results for ``a or b``.
            right = self.parse_expr()
            left = left or right
        return left

    def parse_term(self) -> bool:
        left = self.parse_factor()
        while self._peek() == "and":
            self._eat("and")
            right = self.parse_term()
            left = left and right
        return left

    def parse_factor(self) -> bool:
        t = self._peek()
        if t == "not":
            self._eat("not")
            return not self.parse_factor()
        if t == "(":
            self._eat("(")
            v = self.parse_expr()
            self._eat(")")
            return v
        if t is None:
            raise IfFeatureEvalError("unexpected end of expression")
        self._eat()
        return feature_is_supported(self._ctx, self._enabled, t)

    def at_end(self) -> bool:
        return self._i >= len(self._toks)


def evaluate_if_feature_expression(
    expr: str,
    ctx_module: YangModule,
    enabled_by_module: Mapping[str, AbstractSet[str]],
) -> bool:
    """
    Evaluate a single ``if-feature`` argument string.

    On malformed syntax, returns ``False`` (strict: node is inactive).
    """
    expr = expr.strip()
    if not expr:
        return False
    try:
        p = _IfFeatureParser(_tokenize(expr), ctx_module, enabled_by_module)
        out = p.parse_expr()
        if not p.at_end():
            return False
        return out
    except IfFeatureEvalError:
        return False


def stmt_if_features_satisfied(
    if_features: List[str],
    ctx_module: YangModule,
    enabled_by_module: Mapping[str, AbstractSet[str]],
) -> bool:
    """True iff every ``if-feature`` substatement on the node passes (AND)."""
    if not if_features:
        return True
    return all(
        evaluate_if_feature_expression(e, ctx_module, enabled_by_module)
        for e in if_features
    )


def _prune_per_feature_if_features(
    modules: List[YangModule],
    enabled: Dict[str, AbstractSet[str]],
) -> Dict[str, AbstractSet[str]]:
    """RFC 7950 §7.20.1: drop feature names whose own ``if-feature`` substatements fail."""
    mutable: Dict[str, Set[str]] = {m.name: set(enabled.get(m.name, ())) for m in modules}
    changed = True
    while changed:
        changed = False
        frozen = {mn: frozenset(s) for mn, s in mutable.items()}
        for m in modules:
            for fname in list(mutable[m.name]):
                reqs = m.feature_if_features.get(fname)
                if not reqs:
                    continue
                if not stmt_if_features_satisfied(reqs, m, frozen):
                    mutable[m.name].discard(fname)
                    changed = True
    return {k: frozenset(v) for k, v in mutable.items()}

File: ts/src/validator/if-feature-eval.ts
/**
 * YANG 1.1 if-feature boolean expressions (RFC 7950 §7.20.2, §14).
 * Multiple if-feature substatements on one node are combined with logical AND.
 */

export type ModuleData = Record<string, unknown>;

function stripQuotes(prefix: unknown): string {
  return String(prefix ?? "").replace(/^['"]|['"]$/g, "");
}

function moduleName(m: ModuleData): string {
  return String(m.name ?? "");
}

export function declaredFeatures(m: ModuleData): Set<string> {
  const f = m.features;
  if (f instanceof Set) {
    return new Set(f as Set<string>);
  }
  if (Array.isArray(f)) {
    return new Set(f as string[]);
  }
  return new Set();
}

export function resolvePrefixedModule(ctx: ModuleData, prefix: string): ModuleData | undefined {
  const own = stripQuotes(ctx.prefix);
  if (prefix === own) {
    return ctx;
  }
  const imports = ctx.import_prefixes as Record<string, ModuleData> | undefined;
  const hit = imports?.[prefix];
  if (hit && typeof hit === "object") {
    return hit;
  }
  return undefined;
}

export function reachableModuleData(root: ModuleData): ModuleData[] {
  const out: ModuleData[] = [];
  const seen = new Set<ModuleData>();
  const walk = (m: ModuleData): void => {
    if (seen.has(m)) {
      return;
    }
    seen.add(m);
    out.push(m);
    const im = m.import_prefixes as Record<string, ModuleData> | undefined;
    if (!im) {
      return;
    }
    for (const v of Object.values(im)) {
      if (v && typeof v === "object") {
        walk(v);
      }
    }
  };
  walk(root);
  return out;
}

export function featureIsSupported(
  ctxModule: ModuleData,
  enabledByModule: Readonly<Record<string, ReadonlySet<string>>>,
  ref: string
): boolean {
  let mod: ModuleData;
  let fname: string;
  const idx = ref.indexOf(":");
  if (idx !== -1) {
    const pref = ref.slice(0, idx);
    fname = ref.slice(idx + 1);
    const resolved = resolvePrefixedModule(ctxModule, pref);
    if (!resolved) {
      return false;
    }
    mod = resolved;
  } else {
    mod = ctxModule;
    fname = ref;
  }
  if (!declaredFeatures(mod).has(fname)) {
    return false;
  }
  const enabled = enabledByModule[moduleName(mod)];
  if (!enabled) {
    return false;
  }
  return enabled.has(fname);
}

function tokenize(expr: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  const n = expr.length;
  while (i < n) {
    const c = expr[i]!;
    if (/\s/.test(c)) {
      i += 1;
      continue;
    }
    if (c === "(" || c === ")") {
      tokens.push(c);
      i += 1;
      continue;
    }
    let j = i;
    while (j < n && !/\s/.test(expr[j]!) && expr[j] !== "(" && expr[j] !== ")") {
      j += 1;
    }
    tokens.push(expr.slice(i, j));
    i = j;
  }
  return tokens;
}

class IfFeatureParser {
  private i = 0;

  constructor(
    private readonly toks: string[],
    private readonly ctx: ModuleData,
    private readonly enabled: Readonly<Record<string, ReadonlySet<string>>>
  ) {}

  private peek(): string | undefined {
    return this.toks[this.i];
  }

  private eat(expected?: string): string {
    const t = this.peek();
    if (t === undefined) {
      throw new Error("unexpected end of expression");
    }
    if (expected !== undefined && t !== expected) {
      throw new Error(`expected ${expected}, got ${t}`);
    }
    this.i += 1;
    return t;
  }

  parseExpr(): boolean {
    let left = this.parseTerm();
    while (this.peek() === "or") {
      this.eat("or");
      const right = this.parseExpr();
      left = left || right;
    }
    return left;
  }

  parseTerm(): boolean {
    let left = this.parseFactor();
    while (this.peek() === "and") {
      this.eat("and");
      const right = this.parseTerm();
      left = left && right;
    }
    return left;
  }

  parseFactor(): boolean {
    const t = this.peek();
    if (t === "not") {
      this.eat("not");
      return !this.parseFactor();
    }
    if (t === "(") {
      this.eat("(");
      const v = this.parseExpr();
      this.eat(")");
      return v;
    }
    if (t === undefined) {
      throw new Error("unexpected end of expression");
    }
    this.eat();
    return featureIsSupported(this.ctx, this.enabled, t);
  }

  atEnd(): boolean {
    return this.i >= this.toks.length;
  }
}

export function evaluateIfFeatureExpression(
  expr: string,
  ctxModule: ModuleData,
  enabledByModule: Readonly<Record<string, ReadonlySet<string>>>
): boolean {
  const trimmed = expr.trim();
  if (!trimmed) {
    return false;
  }
  try {
    const p = new IfFeatureParser(tokenize(trimmed), ctxModule, enabledByModule);
    const out = p.parseExpr();
    if (!p.atEnd()) {
      return false;
    }
    return out;
  } catch {
    return false;
  }
}

export function stmtIfFeaturesSatisfied(
  ifFeatures: string[] | undefined,
  ctxModule: ModuleData,
  enabledByModule: Readonly<Record<string, ReadonlySet<string>>>
): boolean {
  if (!ifFeatures || ifFeatures.length === 0) {
    return true;
  }
  return ifFeatures.every((e) => evaluateIfFeatureExpression(e, ctxModule, enabledByModule));
}

function normalizeOverride(
  override: ReadonlyMap<string, ReadonlySet<string>> | Record<string, ReadonlySet<string>> | null | undefined
): Map<string, ReadonlySet<string>> | null {
  if (override == null) {
    return null;
  }
  if (override instanceof Map) {
    return new Map(override);
  }
  return new Map(Object.entries(override));
}

function prunePerFeatureIfFeatures(
  modules: ModuleData[],
  enabled: Record<string, Set<string>>
): Record<string, ReadonlySet<string>> {
  let changed = true;
  while (changed) {
    changed = false;
    const frozen: Record<string, ReadonlySet<string>> = {};
    for (const [mn, s] of Object.entries(enabled)) {
      frozen[mn] = new Set(s);
    }
    for (const m of modules) {
      const mn = moduleName(m);
      const mutable = enabled[mn];
      if (!mutable) {
        continue;
      }
      const fif = m.feature_if_features as Record<string, string[]> | undefined;
      for (const fname of [...mutable]) {
        const reqs = fif?.[fname];
        if (!reqs?.length) {
          continue;
        }
        if (!stmtIfFeaturesSatisfied(reqs, m, frozen)) {
          mutable.delete(fname);
          changed = true;
        }
      }
    }
  }
  const out: Record<string, ReadonlySet<string>> = {};
  for (const [k, v] of Object.entries(enabled)) {
    out[k] = new Set(v);
  }
  return out;
}

/**
 * For each reachable module, the set of feature names that are enabled.
 * - override is null/undefined: every declared feature is enabled.
 * - override[moduleName] present: only those names are enabled for that module.
 * - Module not listed: all declared features enabled for that module.
 */
export function buildEnabledFeaturesMap(
  root: ModuleData,
  override: ReadonlyMap<string, ReadonlySet<string>> | Record<string, ReadonlySet<string>> | null | undefined
): Record<string, ReadonlySet<string>> {
  const modules = reachableModuleData(root);
  const ov = normalizeOverride(override);
  const enabled: Record<string, Set<string>> = {};
  for (const m of modules) {
    const mn = moduleName(m);
    const declared = declaredFeatures(m);
    if (ov === null || !ov.has(mn)) {
      enabled[mn] = new Set(declared);
    } else {
      enabled[mn] = new Set(ov.get(mn) ?? []);
    }
  }
  return prunePerFeatureIfFeatures(modules, enabled);
}
